dsa-solutions/lc-solutions/0000-0099/0004-Median-of-two-Sorted-Array.md

```diff
@@ -0,0 +1,137 @@
+   - If the total number of elements is even, return the average of the two middle elements.
+
+**Python Implementation:**
+```python
+def findMedianSortedArrays(nums1, nums2):
+    merged = []
+    i = j = 0
+    while i < len(nums1) and j < len(nums2):
+        if nums1[i] < nums2[j]:
+            merged.append(nums1[i])
+            i += 1
+        else:
+            merged.append(nums2[j])
+            j += 1
+    merged.extend(nums1[i:])
+    merged.extend(nums2[j:])
+    
+    n = len(merged)
+    if n % 2 == 1:
+        return merged[n // 2]
+    else:
+        return (merged[n // 2 - 1] + merged[n // 2]) / 2.0
+```
+
+#### Approach 2: Binary Search (O(log(min(m, n))))
+
+**Intuition:**
+To achieve O(log(min(m, n))) complexity, use binary search on the smaller array. The goal is to find a partition where the left half of both arrays combined is less than or equal to the right half.
+
+**Algorithm:**
+1. Ensure `nums1` is the smaller array.
+2. Define `imin` and `imax` for binary search on `nums1`.
+3. Calculate `partitionX` and `partitionY` such that:
+   - `partitionX + partitionY = (m + n + 1) // 2`
+4. Check conditions to find the correct partition:
+   - If `maxLeftX <= minRightY` and `maxLeftY <= minRightX`, you've found the correct partition.
+   - If `maxLeftX > minRightY`, move `imax` to the left.
+   - If `maxLeftY > minRightX`, move `imin` to the right.
+5. Calculate the median based on the combined lengths of the arrays.
+
+**Python Implementation:**
+```python
+def findMedianSortedArrays(nums1, nums2):
+    if len(nums1) > len(nums2):
+        nums1, nums2 = nums2, nums1
+    
+    m, n = len(nums1), len(nums2)
+    imin, imax, half_len = 0, m, (m + n + 1) // 2
+    
+    while imin <= imax:
+        partitionX = (imin + imax) // 2
+        partitionY = half_len - partitionX
+        
+        maxLeftX = float('-inf') if partitionX == 0 else nums1[partitionX - 1]
+        minRightX = float('inf') if partitionX == m else nums1[partitionX]
+        
+        maxLeftY = float('-inf') if partitionY == 0 else nums2[partitionY - 1]
+        minRightY = float('inf') if partitionY == n else nums2[partitionY]
+        
+        if maxLeftX <= minRightY and maxLeftY <= minRightX:
+            if (m + n) % 2 == 1:
+                return max(maxLeftX, maxLeftY)
+            else:
+                return (max(maxLeftX, maxLeftY) + min(minRightX, minRightY)) / 2.0
+        elif maxLeftX > minRightY:
+            imax = partitionX - 1
+        else:
+            imin = partitionX + 1
+```
+
+### Complexity Analysis
+
+- **Time Complexity:**
+  - Approach 1: O(m + n) because it involves merging both arrays into one sorted array.
+  - Approach 2: O(log(min(m, n))) because it performs binary search on the smaller array.
+
+- **Space Complexity:**
+  - Approach 1: O(m + n) due to the additional space needed for the merged array.
+  - Approach 2: O(1) because it uses only a constant amount of additional space.
```

Comment:
replace O(1) to `$$O(1)$$`

Reply:
And you change for every files or codes for best practice

Reply:
i think you are trying to say
to replace with $O(1)$ beacuyse 2 doller sign is changing it's alignment